stan/math/prim/mat/fun/gp_matern52_cov.hpp

```diff
@@ -0,0 +1,273 @@
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern52_cov", "x", x[n]);
+
+  check_positive_finite("gp_matern52_cov", "marginal variance", sigma);
+  check_positive_finite("gp_matern52_cov", "length-scale", length_scale);
+
+  Eigen::Matrix<typename stan::return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5_inv_l = pow(5.0, 0.5) / length_scale;
+  T_l inv_l_sq_5_3 = 5.0 / (3.0 * square(length_scale));
+  T_l neg_root_5_inv_l = -1.0 * pow(5.0, 0.5) / length_scale;
+
+  for (size_t i = 0; i < (x_size - 1); ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      cov(i, j) = sigma_sq
+                  * (1.0 + root_5_inv_l * sqrt(squared_distance(x[i], x[j]))
+                     + inv_l_sq_5_3 * squared_distance(x[i], x[j]))
+                  * exp(neg_root_5_inv_l * sqrt(squared_distance(x[i], x[j])));
+      cov(j, i) = cov(i, j);
+    }
+  }
+  cov(x_size - 1, x_size - 1) = sigma_sq;
+  return cov;
+}
+
+/** Returns a Matern 5/2 covariance matrix with one input vector
+ *  with automatic relevance determination (ARD) priors.
+ *
+ * \f[ k(x, x') = \sigma^2(1 +
+ *  \sqrt{5}\sum_{k=1}^{K} sqrt{\frac{d(x, x')^2}{l_k^2}})
+ *  exp(-\frac{5}{3}(\sum_{k=1}^K{\frac{d(x, x')^2}{l_k}})^2)
+ * \f]
+ *
+ * where \f$ d(x, x') \f$ is the euclidean distance.
+ *
+ * @param x std::vector of elements that can be used in stan::math::distance
+ * @param length_scale length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ *
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename stan::return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern52_cov(const std::vector<T_x> &x, const T_s &sigma,
+                const std::vector<T_l> &length_scale) {
+  using stan::math::sqrt;
+  using stan::math::square;
+  using stan::math::squared_distance;
+  using std::exp;
+  using std::pow;
+
+  size_t x_size = x.size();
+  size_t l_size = length_scale.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern52_cov", "x", x[n]);
+
+  check_positive_finite("gp_matern52_cov", "marginal variance", sigma);
+  check_positive_finite("gp_matern52_cov", "length-scale", length_scale);
+
+  Eigen::Matrix<typename stan::return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5 = pow(5.0, 0.5);
+  T_l five_thirds = 5.0 / 3.0;
+  T_l neg_root_5 = -1.0 * pow(5.0, 0.5);
+  T_l temp;
+
+  for (size_t i = 0; i < (x_size - 1); ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      temp = 0;
```

Comment:
Shift the declaration of `temp` to here.

Reply:
We should be able to move the computation of `temp` to outside these for loops and reuse it.

Reply:
can't do this, needs to reset for each element in the matrix, otherwise it would be a cumulative ARD sum of all elements in the matrix, which is invalid... needs to be for each element in the covariance matrix. everything I'm thinking amounts to more steps than this..

Reply:
I don't follow. It looks like this term is used `n * (n - 1) / 2` times (pseudocode): `temp = sum(1 / length_scale[1 : l_size]^2)`

I would think that computation could go outside the loop. Did I miss something?

Reply:
You are correct

Reply:
you're correct on the implementation/optimization, but I'm try to convince myself that this is according to as people define ARD: "seperate length scale for each argument" and in the matlab/GPML document below, as they are index by `m` as well (or k in my implementation)